src/comm.cc

```diff
@@ -471,6 +472,20 @@ comm_apply_flags(int new_socket,
         if ( addr.isNoAddr() )
             debugs(5,0,"CRITICAL: Squid is attempting to bind() port " << addr << "!!");
 
+#if defined(SO_REUSEPORT)
+        if (flags & COMM_REUSEPORT) {
+            int on = 1;
+            if (setsockopt(new_socket, SOL_SOCKET, SO_REUSEPORT, reinterpret_cast<char*>(&on), sizeof(on)) < 0) {
+                const auto savedErrno = errno;
+                const auto errorMessage = ToSBuf("cannot enable SO_REUSEPORT socket option when binding to ",
+                                                 addr, ": ", xstrerr(savedErrno));
+                if (reconfiguring)
+                    debugs(5, DBG_IMPORTANT, errorMessage);
+                else
+                    throw TexcHere(errorMessage);
```

Comment:
I don't think this may be caused by an admin mistake. We get here after setsockopt() generally fails (we had already checked SO_REUSEPORT availability in the system). I am not sure that we should abort Squid after reconfiguring if it can go on with unshared ports. However, in this case, Squid with more than one worker would abort when attempting to open (non-shared due to the error) port in the second worker.

Reply:
> If this admin mistake is safe to ignore on reconfigure surely it is also possible to continue running on the testing parse/check cases as well 

The false precondition in this logic invalidates the conclusions: It is _not_ "safe" to ignore this error on reconfigure. We are "ignoring" it not because it is safe to do so, but because Squid reconfiguration is currently broken at a much higher level -- Squid should detect and reject bad configurations (as a whole) during reconfiguration but does not. That lack of proper reconfiguration support at the higher level forces us to select between these two evils:

* kill running Squid on reconfigure because something went wrong
* report reconfiguration errors but salvage the situation if at all possible

For the feature like this one, the second evil is the lesser one.

> IMO this should [...] always throw

Yes, this code should always throw, but the higher level code is not ready for that yet. When it is ready, it would be trivial to find this code and remove the `if` statement, leaving just its throwing part.

Reply:
The more you are explaining the worse this if-statement is sounding.

The premise behind the need for this feature is that it gets used by admin when the loads are so high that a single worker (or rather CPU) is not capable of handling traffic without some form of hardware/VM issues. On smaller traffic loads it is not useful.

The failure of this SO_* attempt causing other workers to exit will leave just one worker in charge of all traffic hitting the machine. Exactly the opposite of what is required.

That is a recipe for disaster with worst-case also being the use-case where this option is most needed. Which IMO is a clear case for ensuring that a full exit happens as soon as the issue is detected. Especially important that happens for reconfigure/startup/restart. With the current state of the throw handling, that means this code _right now_ will need to use fatal().

Reply:
> The premise behind the need for this feature is that it gets used by admin when the loads are so high that a single worker (or rather CPU) is not capable of handling traffic without some form of hardware/VM issues. On smaller traffic loads it is not useful.

You seem to be talking about SMP in general rather than SO_REUSEPORT in particular, and your implication that SMP (or SO_REUSEPORT) is only useful when a single CPU core is overloaded is incorrect.

SO_REUSEPORT feature is designed for SMP environments. Reducing CPU skew is useful in many such environments, including but not limited to those with overloaded CPU cores.

None of that really matters for this change request AFAICT, but I did not want to leave that misunderstanding uncorrected.

> The failure of this SO_* attempt causing other workers to exit will leave just one worker in charge of all traffic hitting the machine.

Squid should die on startup failures. Where the whole instance death is not supported, each affected worker should die. That is the only sane approach I can think of. After years of fighting, I think (hope?) we agree that new features should follow this approach. This PR code follows this approach.

> Which IMO is a clear case for ensuring that a full exit happens as soon as the issue is detected.

Agreed.

> Especially important that happens for reconfigure/startup/restart. 

Except "reconfigure" -- that case is currently special because:

* many deployments reconfigure _live_ Squids and are not ready to handle such deaths;
* many deployments reconfigure Squids _automatically_, without a human present;
* long-term, Squid should reject invalid configurations (rather than apply them, killing the running instance).

As I said in another comment, we have to pick the lesser of the two evils when reconfiguring until that long-term solution becomes viable.

> With the current state of the throw handling, that means this code _right now_ will need to use fatal().

IIRC, the current effect of throw and fatal are the same in this code -- both result in worker death. @eduard-bagdasaryan, if I am wrong, please replace throw with fatal().

Reply:
> the current effect of throw and fatal are the same in this code

Yes, the exception is caught in SquidMainSafe() and the worker terminates with EXIT_FAILURE.
Leaving as is.

Reply:
@rousskov have you read those "require reading" documents? it is directly stated that what this option does is swap the OS algorithm from LIFO to FIFO listener signalling on accept(). That it was added as a workaround for the cases where one CPU gets overloaded with traffic by the OS using LIFO algorithm.

SMP is only relevant by being the use-case where the problem shows up worst. Leaving the kid1 (being first that is the one which gets no bind() failure) to face the overload alone is counter-productive. Better to have them all exit quickly and have the admin fix things than wait for overload to occur, or worse the "dying too fast" ambiguous error.

Reply:
> have you read those "require reading" documents

s/required/recommended/, but yes, I have read them.

> it is directly stated that what this option does is swap the OS algorithm from LIFO to FIFO listener signalling on accept().

Where do those documents say that? It sounds like you are confusing blocking accept and/or ePoll accept for SO_REUSEPORT. Roughly speaking:

* blocking accept: A FIFO queue of workers
* ePoll accept: A LIFO queue of workers (this is what SMP Squid usually gets on Linux today)
* SO_REUSEPORT: worker selected based on a connection address hash

> SMP is only relevant...

It sounds like you also misunderstood _what_ is being ordered in those FIFO/LIFO queues. It is not connection requests. It is _workers_. With just one worker (i.e. no SMP), there is no difference between the three accept mechanisms at all! Needless to say, after the worker assignment, all assigned connections are accepted (by the assigned worker) in FIFO order of their arrival.

> Better to have them all exit quickly and have the admin fix things

Yes, we [still] agree on that, so I am not sure why you are restating this. The current PR code is the best in-scope approximation of that "have them all exit quickly" idea AFAICT -- in the unlikely event of SO_REUSEPORT setting failure, each failed worker will exit as quickly as it can.

Reply:
> > the current effect of throw and fatal are the same in this code
> 
> Yes, the exception is caught in SquidMainSafe() and the worker terminates with EXIT_FAILURE.
> Leaving as is.

Resolves the review item. I am dropping the side discussion on architecture design for now.